hooks/cascading-scans/hook/reverse-matches.ts

```diff
@@ -0,0 +1,201 @@
+    let mapped = Mustache.render(value, {
+      $: {
+        ...scopeLimiterAliases
+      }
+    });
+    if (mapped == "") {
+      mapped = value;
+    }
+    let rendered = Mustache.render(mapped, finding);
+    return [rendered, rendered != ""];
+  }
+
+  return [
+    scopeLimiter.allOf !== undefined ? scopeLimiter.allOf.every(validateRequirement) : true,
+    scopeLimiter.anyOf !== undefined ? scopeLimiter.anyOf.some(validateRequirement) : true,
+    scopeLimiter.noneOf !== undefined ? !scopeLimiter.noneOf.some(validateRequirement) : true,
+  ].every(entry => entry === true);
+}
+
+interface Operands {
+  lhs: string,
+  rhs: string[],
+}
+
+interface OperatorFunctions {
+  operator: (operands: Operands) => boolean,
+  validator: (operands: Operands) => void,
+}
+
+const defaultValidator: OperatorFunctions["validator"] = props => validate(props, false, false);
+
+const operatorFunctions: { [key in ScopeLimiterRequirementOperator]: OperatorFunctions } = {
+  [ScopeLimiterRequirementOperator.In]: {
+    operator: operatorIn,
+    validator: defaultValidator,
+  },
+  [ScopeLimiterRequirementOperator.NotIn]: {
+    operator: props => !operatorIn(props),
+    validator: defaultValidator,
+  },
+  [ScopeLimiterRequirementOperator.Exists]: {
+    operator: operatorExists,
+    validator: props => validate(props, true, true),
+  },
+  [ScopeLimiterRequirementOperator.DoesNotExist]: {
+    operator: props => !operatorExists(props),
+    validator: props => validate(props, true, true),
+  },
+  [ScopeLimiterRequirementOperator.Contains]: {
+    operator: operatorContains,
+    validator: defaultValidator,
+  },
+  [ScopeLimiterRequirementOperator.DoesNotContain]: {
+    operator: props => !operatorContains(props),
+    validator: defaultValidator,
+  },
+  [ScopeLimiterRequirementOperator.InCIDR]: {
+    operator: operatorInCIDR,
+    validator: defaultValidator,
+  },
+  [ScopeLimiterRequirementOperator.NotInCIDR]: {
+    operator: props => !operatorInCIDR(props),
+    validator: defaultValidator,
+  },
+  [ScopeLimiterRequirementOperator.SubdomainOf]: {
+    operator: operatorSubdomainOf,
+    validator: defaultValidator,
+  },
+  [ScopeLimiterRequirementOperator.NotSubdomainOf]: {
+    operator: props => !operatorSubdomainOf(props),
+    validator: defaultValidator,
+  },
+}
+
+function validate({lhs, rhs}: Operands, lhsUndefinedAllowed, rhsUndefinedAllowed) {
+  if (!lhsUndefinedAllowed && lhs === undefined) {
+    throw new Error(`annotation may not be undefined`)
+  }
+  if (!rhsUndefinedAllowed && rhs === undefined) {
+    throw new Error(`values may not be undefined`)
+  }
+}
+
+function operatorIn({lhs, rhs}: Operands): boolean {
+  return rhs.includes(lhs);
+}
+function operatorExists({lhs, rhs}: Operands): boolean {
+  return lhs !== undefined;
+}
+
+function operatorContains({lhs, rhs}: Operands): boolean {
+  const valueArray = lhs.split(",");
+  return rhs.every(value => valueArray.includes(value));
+}
```

Comment:
I'm still missing some documentation for what the operators are actually supposed to do, so I can't really tell if this is correct behavior. The way I understand it, this operator receives a list of findings in the `value` key (as does every other), but it interprets the provided scope from the scope annotation value as a comma-separated list of values. It then checks if every finding is represented in the comma-separated list of in-scope targets. Is this correct?

I think I'll need some documentation for the expected behavior of these operators before I can continue reviewing this - having to guess at these things is a recipe for introducing unexpected behavior. This documentation should ideally be both inline in this file, and (as I requested in the docs PR) in the written user documentation.

Reply:
I've updated my docs PR with short explanation for each operator. I'll copy them to source code as soon as you approve them in the docs PR, to prevent double work.

Reply:
Great, thank you. The explanations in the docs PR are fine, feel free to copy-paste them here.

Reply:
0c1d8c9 - Add inline documentation for every operator